Framework: Make create-selector call dependant state getter with arguments

samuelclemens opened the pull request:
Consider the following selector, which goal is to build a comments tree for a given post from a list of comments which are stored in a structure:

```javascript
state.comments[ getCommentParentKey( siteId, postId ) ]  = [...comments...];
```

```javascript
export const getPostCommentItems = ( state, siteId, postId ) => state.comments.items.get( getCommentParentKey( siteId, postId ) );

export const getPostCommentsTree = createSelector(
	( state, siteId, postId ) => {
		const items = getPostCommentItems( state, siteId, postId );
		return items ? buildCommentsTree( items ) : undefined;
	},
	getPostCommentItems
);
```

In order to get the dependent state in an efficient manner, the arguments should be passed to the dependant state getter.
We can't simply return `state.comments`, that will invalidate all of the post's comments trees even if only one comments list is changed.

blowery commented:
cc @aduth as he wrote `create-selector`

aduth commented:
I don't personally have much of a problem with this, though generally it seems simpler to bust at the paths which are accessible only by navigating the reducer keys of the tree.

A few points to make:

- Can we update the documentation for the module to reflect this new behavior? Or should this be a "hidden" feature such that we deemphasize its use?

- You should probably use `calledWithExactly` in your test, not `calledWith`, as if you were to change the call to the selector with `memoizedSelector( state, 1, 2, 3, 4 );`, it would still pass.

- There's a merge conflict, likely related to the new test added with functionality at #3575

samuelclemens commented:
@aduth Thanks for taking a look at this.
I've rebased, adjusted the test and added reference to this in the readme.

I guess for a light computation it does makes sense to bust just at the state's tree keys, but as 
the memoized computation gets more complex imo it's better to allow the fine tuning of the 
defendant state.

aduth commented:
Personally, I think it's sufficient to detail the argument signature of the dependants function in the docs. I don't find the complex example easy to grasp at a glance without a better understanding of the comments tree structure, and would rather not include it.

samuelclemens commented:
Something like that?

aduth commented:
Precisely. Thanks @samuelclemens !

samuelclemens commented:
@aduth Thank you =)
